File: Assets/Code/Logic/UI/CommonItem/View/CommonItemIcon.cs
using UnityEngine;
using UnityEngine.UI;
using Common.Localization;
using Common.ResMgr;
using Logic.Enums;
using Logic.Game.Model;
using Logic.Item.Model;
using Logic.UI.Description.View;
using Logic.Audio.Controller;
using System.Collections.Generic;
using Logic.Hero.Model;
using Logic.Equipment.Model;

namespace Logic.UI.CommonItem.View
{
    public class CommonItemIcon : MonoBehaviour
    {
        public const string PREFAB_PATH = "ui/common/common_item_icon";

        public static CommonItemIcon Create(Transform parent)
        {
            GameObject go = ResMgr.instance.Load<GameObject>(PREFAB_PATH);
            go = GameObject.Instantiate<GameObject>(go);
            if (parent != null)
            {
                go.transform.SetParent(parent, false);
                go.transform.localPosition = Vector3.zero;
            }
            return go.GetComponent<CommonItemIcon>();
        }

        private ItemInfo _itemInfo;

        public ItemInfo ItemInfo
        {
            get
            {
                return _itemInfo;
            }
        }


        public bool isSelect;


        public delegate void OnClickHandler(CommonItemIcon icon);
        public OnClickHandler onClickHandler;

        #region UI components
        public RectTransform rectTransform;
		public Image itemQualityFrameImage;
        public Image itemIconImage;
		public Image pieceMarkImage;
        public GameObject selectMark;
        public Text textCount;
        public Text textName;

		public Image roleTypeIcon;
		public Transform starsRoot;
		public List<Image> starImages;
        #endregion UI components

        private ItemDesButton _itemDesButton;
        void Awake()
        {
            SetSelect(false);
            _itemDesButton = ItemDesButton.Get(gameObject);
        }

        private void RefreshItemIconImage()
        {
            Sprite sprite = ResMgr.instance.Load<Sprite>(ResPath.GetItemIconPath(_itemInfo.itemData.icon));
            if (sprite != null)
            {
                itemIconImage.SetSprite(sprite);
                //itemIconImage.SetNativeSize();
            }
        }

        public void Refresh()
        {
			itemQualityFrameImage.SetSprite( UIUtil.GetItemQualityFrameSprite(_itemInfo.itemData.itemQuality));

            RefreshItemIconImage();

			if (_itemInfo.itemData.type == (int)ITEM_TYPE.HeroPiece)
			{
				PieceData heroPieceData = PieceData.GetPieceDataByID(_itemInfo.itemData.id);
				HeroInfo heroInfo = new HeroInfo(heroPieceData.heroGameResData);
				itemIconImage.SetSprite(ResMgr.instance.LoadSprite(heroInfo.HeadIcon));

				for (int i = 0, count = starImages.Count; i < count; i++)
				{
					if (i < heroInfo.advanceLevel)
						starImages[i].SetSprite(ResMgr.instance.LoadSprite("sprite/main_ui/icon_star"));
					else
						starImages[i].SetSprite(ResMgr.instance.LoadSprite("sprite/main_ui/icon_star2_big_disable"));
					starImages[i].gameObject.SetActive(i < heroInfo.MaxAdvanceLevel);
				}

				pieceMarkImage.gameObject.SetActive(true);
				roleTypeIcon.SetSprite(UIUtil.GetRoleTypeSmallIconSprite(heroInfo.heroData.roleType));
				roleTypeIcon.gameObject.SetActive(true);
//				starsRoot.gameObject.SetActive(_itemInfo.itemData.star > 0);
			}
			else if (_itemInfo.itemData.type == (int)ITEM_TYPE.EquipPiece)
			{
				PieceData pieceData = PieceData.GetPieceDataByID(_itemInfo.itemData.id);
				EquipmentData equipmentData = EquipmentData.GetEquipmentDataByID(pieceData.heroGameResData.id);
				itemIconImage.SetSprite(ResMgr.instance.LoadSprite(ResPath.GetEquipmentIconPath(equipmentData.icon)));

				pieceMarkImage.gameObject.SetActive(true);
				roleTypeIcon.SetSprite(UIUtil.GetRoleTypeSmallIconSprite(equipmentData.equipmentRoleType));
				roleTypeIcon.gameObject.SetActive(true);
			}
			else
			{
				pieceMarkImage.gameObject.SetActive(false);
				roleTypeIcon.gameObject.SetActive(false);
//				starsRoot.gameObject.SetActive(false);
			}
			starsRoot.gameObject.SetActive(false);

            textCount.text = _itemInfo.count.ToString();
            if (_itemInfo.count >= 1)
                ShowCount();
            else
                HideCount();

            textName.text = _itemInfo.itemData.id.ToString();
            textName.gameObject.SetActive(false);
        }

        public void SetItemInfo(ItemInfo itemInfo)
        {
            _itemInfo = itemInfo;
            Refresh();
            _itemDesButton.SetItemInfo(_itemInfo);
        }

		public void SetEnableItemDesButton (bool enable)
		{
			_itemDesButton.enabled = enable;
		}

        public void SetGameResData(GameResData gameResData)
        {
            int id = gameResData.id;

            BaseResType type = gameResData.type;
            if (type != BaseResType.Hero && type != BaseResType.Equipment && type != BaseResType.Item)
            {
                ItemData data = ItemData.GetBasicResItemByType(type);
                if (data != null)
                    id = data.id;
            }
            else
            {

            }
            SetItemInfo(new ItemInfo(0, id, gameResData.count));
        }

        public void SetAsCostResources(GameResData gameResData)
        {
            SetSelect(false);
            textName.gameObject.SetActive(false);
            //Sprite iconSprite = null;
            int requiredCount = 0;
            int ownCount = 0;

            BaseResType type = gameResData.type;
            int id = gameResData.id;
            if (type != BaseResType.Hero && type != BaseResType.Equipment && type != BaseResType.Item)
            {

                ownCount = GameProxy.instance.BaseResourceDictionary[gameResData.type];
                ItemData data = ItemData.GetBasicResItemByType(type);
                if (data != null)
                    id = data.id;
            }
            else
            {
                ItemInfo itemInfo = ItemProxy.instance.GetItemInfoByItemID(gameResData.id);
                if (itemInfo != null)
                {
                    ownCount = itemInfo.count;
                }
            }
            requiredCount = gameResData.count;
            //			if (gameResData.type == BaseResType.Gold
            //			    || gameResData.type == BaseResType.Diamond
            //			    || gameResData.type == BaseResType.Crystal
            //			    || gameResData.type == BaseResType.Honor)
            //			{
            //				iconSprite = ResMgr.instance.Load<Sprite>(ResPath.GetItemIconPath(((int)gameResData.type).ToString()));
            //				requiredCount = gameResData.count;
            //				ownCount = GameProxy.instance.BaseResourceDictionary[gameResData.type];
            //			}
            //			else if (gameResData.type == BaseResType.Item)
            //			{
            //				ItemData itemData = ItemData.GetItemDataByID(gameResData.id);
            //				iconSprite = ResMgr.instance.Load<Sprite>(ResPath.GetItemIconPath(itemData.icon));
            //				requiredCount = gameResData.count;
            //
            //				ItemInfo itemInfo = ItemProxy.instance.GetItemInfoByItemID(gameResData.id);
            //				if (itemInfo != null)
            //				{
            //					ownCount = itemInfo.count;
            //				}
            //			}
            //			itemIconImage.sprite = iconSprite;
            //			itemIconImage.SetNativeSize();
            SetItemInfo(new ItemInfo(0, id, gameResData.count));
            textCount.text = ownCount >= requiredCount ? UIUtil.FormatToGreenText(requiredCount.ToString()) : UIUtil.FormatToRedText(requiredCount.ToString());

        }

		public void SetAsRequireResource (GameResData gameResData)
		{
			SetSelect(false);
			textName.gameObject.SetActive(false);
			int requiredCount = 0;
			int ownCount = 0;

			BaseResType type = gameResData.type;
			int id = gameResData.id;
			if (type != BaseResType.Hero && type != BaseResType.Equipment && type != BaseResType.Item)
			{
				
				ownCount = GameProxy.instance.BaseResourceDictionary[gameResData.type];
				ItemData data = ItemData.GetBasicResItemByType(type);
				if (data != null)
					id = data.id;
			}
			else
			{
				ItemInfo itemInfo = ItemProxy.instance.GetItemInfoByItemID(gameResData.id);
				if (itemInfo != null)
				{
					ownCount = itemInfo.count;
				}
			}
			requiredCount = gameResData.count;
			SetItemInfo(new ItemInfo(0, id, gameResData.count));
			string countString = string.Format("{0}/{1}", ownCount, requiredCount);
			textCount.text = ownCount >= requiredCount ? UIUtil.FormatToGreenText(countString) : UIUtil.FormatToRedText(countString);
		}

        public void SetSelect(bool select)
        {
            isSelect = select;
            selectMark.SetActive(select);
        }

        public void ShowSelect()
        {
            selectMark.SetActive(true);
        }
        public void HideSelect()
        {
            selectMark.SetActive(false);
        }
        public void ShowCount()
        {
            textCount.gameObject.SetActive(true);
        }
        public void HideCount()
        {
            textCount.gameObject.SetActive(false);
        }

        public void SetMiddlePointAsPivot()
        {
            rectTransform.pivot = new Vector2(0.5f, 0.5f);
        }

        public void SetBottomCenterAsPivot()
        {
            rectTransform.pivot = new Vector2(0.5f, 0);
        }

        #region ui event handlers
        public void ClickHandler(CommonItemIcon icon)
        {
            if (onClickHandler != null)
            {
                onClickHandler(icon);
                AudioController.instance.PlayAudio(AudioController.SELECT, false);
            }
        }
        #endregion ui event handlers
    }
}

File: Assets/ToLua/Lua/user/ui/common_icon/common_item_icon.lua
local t = {}
local PREFAB_PATH = 'ui/common/common_item_icon_lua'
t.__index = t



function t.New(parent)
 
  local gameObject = GameObject.Instantiate(Common.ResMgr.ResMgr.instance:Load(PREFAB_PATH))
  
  local o = t.NewByGameObject(gameObject)
  o.transform:SetParent(parent,false)
  
  return o
end

function t.NewByGameObject(gameObject)
  local o = {}
  setmetatable(o,t)
  
  o.transform = gameObject:GetComponent(typeof(Transform))
  
  --callback contains one obj
  o.onClick = void_delegate.New()
  o.isSelect = false
  o:InitComponent()
  return o
end
function t:InitComponent()
  self.itemQualityFrame = self.transform:GetComponent(typeof(Image))
  self.imgIcon = self.transform:Find('img_icon'):GetComponent(typeof(Image))
  self.textCount = self.transform:Find('text_count'):GetComponent(typeof(Text))
  
  local tranImgPieceMark = self.transform:Find('img_piece_mark')
  if tranImgPieceMark ~= nil then
    self.pieceMarkImage = tranImgPieceMark:GetComponent(typeof(Image))
  end
  
  self.tranStarsRoot = self.transform:Find('stars_root')
  self.starImages = {}
  for i = 1, 6 do
    local starImage = self.tranStarsRoot:Find('img_star_'..i):GetComponent(typeof(Image))
    table.insert(self.starImages, starImage)
  end
  
  local tranImgRoleType = self.transform:Find('img_role_type')
  if tranImgRoleType ~= nil then
    self.roleTypeIconImage = tranImgRoleType:GetComponent(typeof(Image))
  end
  
  self.goNewMarkTransform = self.transform:Find('img_new_mark')
  if self.goNewMarkTransform ~= nil then
    self.goNewMarkTransform.gameObject:SetActive(false)
  end
  self.goSelect = self.transform:Find('img_select').gameObject
  self.goSelect:SetActive(false)
  
  self.btn = self.transform:GetComponent(typeof(Button))
  self:ResetListener ()
end
function t:SetActive(active)
  self.transform.gameObject:SetActive(active)
end
function t:SetItemInfo(itemInfo)
  self.itemInfo = itemInfo
  
  self:Refresh()
  self:AddItemDesButton(true)
end

function t:SetGameResData(gameResData)
  local itemInfo = require('ui/item/model/item_info')

  if (gameResData.type ~= BaseResType.Hero and gameResData.type ~= BaseResType.Equipment and gameResData.type ~= BaseResType.Item ) then
      local data = gamemanager.GetData('item_data').GetBasicResItemByType(gameResData.type)
      if data then
        itemInfo = itemInfo.New(0,data.id,gameResData.count)
      end
  else
        itemInfo = itemInfo.New(0,gameResData.id,gameResData.count)
  end
  self:SetItemInfo(itemInfo)
end

function t:SetItemDataId(itemDataId)
   local itemInfo = require('ui/item/model/item_info')
  itemInfo = itemInfo.New(0,itemDataId,0)
  self:SetItemInfo(itemInfo)
end
function t:AddItemDesButton(isLongPress)
  if isLongPress == nil then
    isLongPress = true
  end
  LuaInterface.LuaCsTransfer.GetItemDesButton(self.transform.gameObject, self.itemInfo.itemData.id,isLongPress)
end
function t:Refresh()
  self.itemQualityFrame.sprite = ui_util.GetItemQualityFrameSprite(self.itemInfo.itemData.quality)
  if self.itemInfo.itemData.type == ItemType.HeroPiece then
    local heroPieceData = gamemanager.GetData('piece_data').GetDataById(self.itemInfo.itemData.id)
    local hero_info = require('ui/hero/model/hero_info')
    local heroInfo = hero_info:NewByGameResData(heroPieceData:GetHeroGameResData())
    self.imgIcon.sprite = Common.ResMgr.ResMgr.instance:LoadSprite(heroInfo:HeadIcon())
    
    if self.pieceMarkImage ~= nil then
      self.pieceMarkImage.gameObject:SetActive(true)
    end
    if self.tranStarsRoot ~= nil then
      local spriteStarNormal = Common.ResMgr.ResMgr.instance:LoadSprite('sprite/main_ui/icon_star')
      local spriteStarDefault = Common.ResMgr.ResMgr.instance:LoadSprite('sprite/main_ui/icon_star2_big_disable')

      for k, v in ipairs(self.starImages) do
        if k <= heroInfo.advanceLevel then
          v.sprite = spriteStarNormal
        else
          v.sprite = spriteStarDefault
        end
        v.gameObject:SetActive(k <= heroInfo:MaxAdvanceLevel())
      end
      --self.tranStarsRoot.gameObject:SetActive(self.itemInfo.itemData.star > 0)
      self.tranStarsRoot.gameObject:SetActive(false)
    end
    if self.roleTypeIconImage ~= nil then
      self.roleTypeIconImage.sprite = heroInfo:RoleTypeIconSprite()
      self.roleTypeIconImage.gameObject:SetActive(true)
    end
  elseif self.itemInfo.itemData.type == ItemType.EquipPiece then
    local pieceData = gamemanager.GetData('piece_data').GetDataById(self.itemInfo.itemData.id)
    local equipData = gamemanager.GetData('equip_data').GetDataById(pieceData:GetEquipGameResData().id)
    self.imgIcon.sprite = Common.ResMgr.ResMgr.instance:LoadSprite(equipData:IconPath())
    
    if self.tranStarsRoot ~= nil then
      self.tranStarsRoot.gameObject:SetActive(false)
    end
    if self.pieceMarkImage ~= nil then
      self.pieceMarkImage.gameObject:SetActive(true)
    end
    if self.roleTypeIconImage ~= nil then
      self.roleTypeIconImage.sprite = equipData:RoleTypeIconSprite()
      self.roleTypeIconImage.gameObject:SetActive(true)
    end
  else
    self.imgIcon.sprite = Common.ResMgr.ResMgr.instance:LoadSprite(self.itemInfo:ItemIcon())
    
    if self.pieceMarkImage ~= nil then
      self.pieceMarkImage.gameObject:SetActive(false)
    end
    
    if self.tranStarsRoot ~= nil then
      self.tranStarsRoot.gameObject:SetActive(false)
    end
    
    if self.roleTypeIconImage ~= nil then
      self.roleTypeIconImage.gameObject:SetActive(false)
    end
  end
  
  local count = self.itemInfo:Count()
  --if count < 1 then count = 1 end
  self.textCount.text = tostring(count)
   if count >= 1 then
     self.textCount.gameObject:SetActive(true)
   else
     self.textCount.gameObject:SetActive(false)
   end
end
function t:SetSelect(isSelect)
  self.goSelect:SetActive(isSelect)
  self.isSelect = isSelect
end

function t:ShowCount(isShow)
  self.textCount.gameObject:SetActive(isShow)
end
function t:ShowNewMark(isNew)
  if self.goNewMarkTransform ~= nil then
    self.goNewMarkTransform.gameObject:SetActive(isNew)
  end
end
function t:SetCount(countString,color)
  self:ShowCount(true)
  self.textCount.text = countString
  if color ~= nil then
    self.textCount.color = color
  end
end

function t:ResetListener ()
  self.btn.onClick:RemoveAllListeners()
  self.btn.onClick:AddListener(
    function()
       --print('instanceId:',self.itemInfo.instanceId,'modelId:',self.itemInfo.itemData.id)
      self.onClick:InvokeOneParam(self)
    end)
end

return t